plugins/volume/persistent.go

```diff
@@ -0,0 +1,145 @@
+package volume
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/pborman/uuid"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/plugins/extpoints"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+)
+
+type persistentVolumePlugin struct {
+	plugins.PluginBase
+	mutex       sync.Mutex
+	volumes     map[string][]*persistentVolume
+	environment *runtime.Environment
+	engine      engines.Engine
+	log         *logrus.Entry
+}
+
+// NewPersistentVolumePlugin is a single plugin that will live between task runs.
+// A list of volumes are maintained so that they can be used between task runs
+// so long as the volume is not currently claimed by a task.
+func NewPersistentVolumePlugin(opts extpoints.PluginOptions) (volumePlugin, error) {
+	return &persistentVolumePlugin{
+		environment: opts.Environment,
+		engine:      opts.Engine,
+		log:         opts.Log.WithField("Plugin", "persistent"),
+		volumes:     make(map[string][]*persistentVolume),
+	}, nil
+}
+
+// newVolume will be called by the volume task plugin during the Prepare stage.
+// If a previously used volume is available (i.e. not claimed by a task), then
+// the volume will be marked as claimed by this task and reused.  If no available
+// volume is found, a new volume will be requested from the engine and saved within
+// the list of persistent volumes.
+func (pv *persistentVolumePlugin) newVolume(opts *volumeOptions) (volume, error) {
+	if v := pv.claimAvailableVolume(opts); v != nil {
+		return v, nil
+	}
+
+	// Create a volume since an available one doesn't exist
+	return pv.createVolume(opts)
+}
+
+// claimAvailableVolume will claim and return a volume is one is found that is not
+// in use currently by a task.
+func (pv *persistentVolumePlugin) claimAvailableVolume(opts *volumeOptions) *persistentVolume {
+	pv.mutex.Lock()
+	defer pv.mutex.Unlock()
+	volumes, _ := pv.volumes[opts.name]
+	for _, v := range volumes {
```

Comment:
yea if we want to get clever later with how we search we can, I don't expect any of this to be carved in stone.  I can't imagine we would have more than a couple of volumes per name because most of our workers are running with only 1 task at a time anyways.